File: scripts/share-e2e/seed.py
"""Seed a fresh sharing fixture: two users, a marked-up paper, both links.

Run before `run.mjs`. Every run makes a new paper and a new recipient, so the
suite can assert that the recipient has not got the paper yet — which it could
not do twice against the same one.

    python3 scripts/share-e2e/seed.py

PAPOL_BASE     where the development server is (default http://127.0.0.1:8010)
PAPOL_E2E_PDF  an upload already in the backend's uploads directory
PAPOL_E2E_FIXTURE  where to write the fixture (default: a file in the temp dir)
"""
import json
import os
import tempfile
import urllib.error
import urllib.request

BASE = os.environ.get("PAPOL_BASE", "http://127.0.0.1:8010")
PASSWORD = "papol-test-pw"
PDF = os.environ.get(
    "PAPOL_E2E_PDF",
    "6e4e3411984f3edf99dbfe8b941cb5e8a321379ff0cae6ae5c1f592ad8882ca8.pdf",
)
FIXTURE = os.environ.get(
    "PAPOL_E2E_FIXTURE", os.path.join(tempfile.gettempdir(), "papol-share-e2e.json"))


def call(method, path, token=None, body=None):
    data = json.dumps(body).encode() if body is not None else None
    request = urllib.request.Request(BASE + path, data=data, method=method)
    if data is not None:
        request.add_header("Content-Type", "application/json")
    if token:
        request.add_header("Authorization", f"Bearer {token}")
    try:
        with urllib.request.urlopen(request) as response:
            text = response.read().decode()
            return response.status, (json.loads(text) if text.strip() else None)
    except urllib.error.HTTPError as error:
        text = error.read().decode()
        return error.code, (json.loads(text) if text.strip() else None)
    except urllib.error.URLError as error:
        raise SystemExit(
            f"No Papol at {BASE} ({error.reason}). Start one with "
            f"PAPOL_DEV_PORT=8010 ./deploy.sh dev") from error


def account(email, name):
    """Register, or sign in if this run has been made before."""
    status, out = call("POST", "/api/auth/register",
                       body={"email": email, "display_name": name, "password": PASSWORD})
    if status != 200:
        status, out = call("POST", "/api/auth/login",
                           body={"email": email, "password": PASSWORD})
        if status != 200:
            raise SystemExit(f"could not sign {email} in: {out}")
    return out["token"], out["user"]["uuid"], out["user"]["display_name"]


suffix = os.urandom(3).hex()
sharer_token, sharer_uuid, sharer_name = account("sharer@papol.test", "Alice Sharer")
user_token, user_uuid, user_name = account(f"user-{suffix}@papol.test", "Bob User")

title = f"A Reading Worth Handing Over {suffix}"
status, paper = call("POST", "/api/papers", token=sharer_token, body={
    "title": title, "file_path": PDF, "authors": "A. Sharer", "year": 2026})
if status != 200:
    raise SystemExit(
        f"could not make a paper from {PDF}: {paper}\n"
        "Name an upload this server already holds with PAPOL_E2E_PDF.")
paper_uuid, edition_uuid = paper["uuid"], paper["edition_uuid"]

# A note and a stroke, so the suite can tell a rich link from a lean one by
# what reaches the page rather than by what the API says.
note = f"Alice's note {suffix} — this should reach whoever follows the link"
assert call("POST", f"/api/papers/{paper_uuid}/annotations", token=sharer_token, body={
    "kind": "note", "edition_uuid": edition_uuid, "page": 1, "content": note,
    "body": {"anchor": {"type": "point", "x": 0.3, "y": 0.4}}})[0] == 200
assert call("POST", f"/api/papers/{paper_uuid}/annotations", token=sharer_token, body={
    "kind": "ink", "edition_uuid": edition_uuid, "page": 1,
    "body": {"points": [{"x": 0.15, "y": 0.25}, {"x": 0.55, "y": 0.28}],
             "color": "#b3923d", "width": 0.006, "opacity": 0.9, "shape": "round"}})[0] == 200

_, rich = call("POST", f"/api/papers/{paper_uuid}/sharable", token=sharer_token,
               body={"include_marks": True})
_, lean = call("POST", f"/api/papers/{paper_uuid}/sharable", token=sharer_token,
               body={"include_marks": False})

with open(FIXTURE, "w") as written:
    json.dump({
        "base": BASE, "title": title, "note": note,
        "sharer": {"token": sharer_token, "uuid": sharer_uuid, "name": sharer_name},
        "user": {"token": user_token, "uuid": user_uuid, "name": user_name},
        "paper_uuid": paper_uuid, "edition_uuid": edition_uuid,
        "rich": rich["uuid"], "lean": lean["uuid"],
        "rich_url": f"{BASE}/viewer/?share={rich['uuid']}",
        "lean_url": f"{BASE}/viewer/?share={lean['uuid']}",
    }, written, indent=1)

print(f"seeded {title}")
print(f"  rich {rich['uuid'][:8]}  lean {lean['uuid'][:8]}  recipient {user_name}")
print(f"  fixture at {FIXTURE}")

File: scripts/share-e2e/seed.mjs
// Seed a fresh sharing fixture: two users, a marked-up paper, both links.
//
// Run before `run.mjs`. Every run makes a new paper and a new recipient, so
// the suite can assert that the recipient has not got the paper yet — which
// it could not do twice against the same one.
//
//     node scripts/share-e2e/seed.mjs
//
// PAPOL_BASE         where Papol is (default http://127.0.0.1:5173, the
//                    frontend `./deploy.sh dev` serves, with the Worker
//                    behind it; in CI, the Worker on the assembled site)
// PAPOL_E2E_PDF      a PDF file to upload (default: a fresh page is
//                    written for this run)
// PAPOL_E2E_FIXTURE  where to write the fixture (default: a file in the
//                    temp dir)

import { readFileSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { createHash, randomBytes } from 'node:crypto';

const BASE = (process.env.PAPOL_BASE || 'http://127.0.0.1:5173').replace(/\/$/, '');
const PASSWORD = 'papol-test-pw';
const suffix = randomBytes(3).toString('hex');
const FIXTURE = process.env.PAPOL_E2E_FIXTURE || join(tmpdir(), 'papol-share-e2e.json');

async function call(method, path, { token, body, form } = {}) {
  const headers = {};
  if (token) headers.Authorization = `Bearer ${token}`;
  if (body !== undefined) headers['Content-Type'] = 'application/json';
  let response;
  try {
    response = await fetch(BASE + path, {
      method,
      headers,
      body: form ?? (body === undefined ? undefined : JSON.stringify(body)),
    });
  } catch (error) {
    console.error(`No Papol at ${BASE} (${error.cause?.code || error.message}). Start one with ./deploy.sh dev`);
    process.exit(2);
  }
  const text = await response.text();
  return [response.status, text.trim() ? JSON.parse(text) : null];
}

// Register, or sign in if this run has been made before.
async function account(email, name) {
  let [status, out] = await call('POST', '/api/auth/register', {
    body: { email, display_name: name, password: PASSWORD },
  });
  if (status !== 200) {
    [status, out] = await call('POST', '/api/auth/login', { body: { email, password: PASSWORD } });
    if (status !== 200) throw new Error(`could not sign ${email} in: ${JSON.stringify(out)}`);
  }
  return { token: out.token, uuid: out.user.uuid, name: out.user.display_name };
}

// One page with this run's suffix printed on it. New words are new bytes,
// and the digest of the bytes is the paper's identity — which is what makes
// the paper genuinely new each run rather than a reused one.
function freshPdf(text) {
  const stream = `BT /F1 18 Tf 72 720 Td (${text.replace(/[\\()]/g, (c) => `\\${c}`)}) Tj ET`;
  const objects = [
    '<< /Type /Catalog /Pages 2 0 R >>',
    '<< /Type /Pages /Kids [3 0 R] /Count 1 >>',
    '<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] /Contents 4 0 R /Resources << /Font << /F1 5 0 R >> >> >>',
    `<< /Length ${Buffer.byteLength(stream)} >>\nstream\n${stream}\nendstream`,
    '<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>',
  ];
  let body = '%PDF-1.4\n';
  const offsets = [];
  objects.forEach((object, index) => {
    offsets.push(Buffer.byteLength(body));
    body += `${index + 1} 0 obj\n${object}\nendobj\n`;
  });
  const xref = Buffer.byteLength(body);
  body += `xref\n0 ${objects.length + 1}\n0000000000 65535 f \n`;
  for (const offset of offsets) body += `${String(offset).padStart(10, '0')} 00000 n \n`;
  body += `trailer\n<< /Size ${objects.length + 1} /Root 1 0 R >>\nstartxref\n${xref}\n%%EOF\n`;
  return Buffer.from(body, 'latin1');
}

const pdf = process.env.PAPOL_E2E_PDF
  ? readFileSync(process.env.PAPOL_E2E_PDF)
  : freshPdf(`A reading worth handing over ${suffix}`);

const sharer = await account('sharer@papol.test', 'Alice Sharer');
const user = await account(`user-${suffix}@papol.test`, 'Bob User');

// The bytes first, into the bucket under their digest by the address the
// Worker gives (its own door, on a local one); then the paper that names them.
const digest = createHash('sha256').update(pdf).digest('hex');
const [addressStatus, address] = await call('POST', '/api/files/upload-address', {
  token: sharer.token, body: { kind: 'paper', sha256: digest, size: pdf.length, name: `e2e-${suffix}.pdf` },
});
if (addressStatus !== 200) throw new Error(`could not get an address for the PDF: ${JSON.stringify(address)}`);
if (!address.stored) {
  const put = await fetch(/^https?:/.test(address.url) ? address.url : BASE + address.url, { method: 'PUT', headers: address.headers, body: pdf });
  if (!put.ok) throw new Error(`could not store the PDF: the bucket answered ${put.status}`);
}
const [uploadStatus, upload] = await call('POST', '/api/papers/uploaded', {
  token: sharer.token, body: { file_path: address.file_path, uploaded_name: `e2e-${suffix}.pdf` },
});
if (uploadStatus !== 202) throw new Error(`could not upload the PDF: ${JSON.stringify(upload)}`);

const title = `A Reading Worth Handing Over ${suffix}`;
const [status, paper] = await call('POST', '/api/papers', {
  token: sharer.token,
  body: { title, file_path: upload.file_path, authors: JSON.stringify(['A. Sharer']), year: 2026 },
});
if (status !== 200) throw new Error(`could not make a paper from ${upload.file_path}: ${JSON.stringify(paper)}`);
const paperSha256 = paper.sha256;
// The service answers to a paper's name, which is half its digest.
const paperName = paperSha256.slice(0, 32);

const annotate = async (body) => {
  const [code, out] = await call('POST', `/api/papers/${paperName}/annotations`, { token: sharer.token, body });
  if (code !== 200) throw new Error(`could not annotate: ${JSON.stringify(out)}`);
};
// A note and a stroke, so the suite can tell a rich link from a lean one by
// what reaches the page rather than by what the API says.
const note = `Alice's note ${suffix} — this should reach whoever follows the link`;
await annotate({
  kind: 'note', page: 1, content: note,
  body: { anchor: { type: 'point', x: 0.3, y: 0.4 } },
});
await annotate({
  kind: 'ink', page: 1,
  body: { points: [{ x: 0.15, y: 0.25 }, { x: 0.55, y: 0.28 }], color: '#b3923d', width: 0.006, opacity: 0.9, shape: 'round' },
});
// And a clip, so the sharer's own reading of the paper has one to draw.
await annotate({
  kind: 'clip', page: 1,
  body: { source: { x: 0.1, y: 0.1, w: 0.25, h: 0.15 }, frame: { x: 0.5, y: 0.45, w: 0.25, h: 0.15 }, floating: false },
});

const [, rich] = await call('POST', `/api/papers/${paperName}/sharable`, { token: sharer.token, body: { include_annotations: true } });
const [, lean] = await call('POST', `/api/papers/${paperName}/sharable`, { token: sharer.token, body: { include_annotations: false } });

writeFileSync(FIXTURE, JSON.stringify({
  base: BASE, title, note,
  sharer, user,
  paper_sha256: paperSha256, paper_name: paperName,
  rich: rich.uuid, lean: lean.uuid,
  rich_url: `${BASE}/viewer/?share=${rich.uuid}`,
  lean_url: `${BASE}/viewer/?share=${lean.uuid}`,
  nook_url: `${BASE}/viewer/?pdf=${paperSha256}`,
}, null, 1));

console.log(`seeded ${title}`);
console.log(`  rich ${rich.uuid.slice(0, 8)}  lean ${lean.uuid.slice(0, 8)}  recipient ${user.name}`);
console.log(`  fixture at ${FIXTURE}`);
